resharper/resharper-unity/src/CSharp/Feature/Services/QuickFixes/InefficientMultidimensionalArrayUsageQuickFix.cs

```diff
@@ -0,0 +1,146 @@
+            {
+                var newType = TypeFactory.CreateArrayType(TypeFactory.CreateArrayType(myType.ElementType, 1, NullableAnnotation.NotNullable),
+                    1, myType.NullableAnnotation);
+                myVariableDeclaration.SetType(newType);
+            }
+
+
+            var factory = CSharpElementFactory.GetInstance(myVariableDeclaration);
+
+            if (myArrayCreationExpression != null)
+            {
+                if (myArrayCreationExpression.Initializer == null)
+                {
+                    ConvertArrayCreationWithoutInitializer(factory);
+                }
+                else
+                {
+                    ConvertArrayCreation(factory);
+                }
+            }
+
+            ConvertUsages(factory);
+
+            return null;
+        }
+
+        private void ConvertArrayCreation(CSharpElementFactory factory)
+        {
+            var currentCreation = myArrayCreationExpression;
+            var pattern = new StringBuilder("new [] {");
+            var args = new List<object>();
+            args.Add(myType.ElementType);
+
+            var elements = currentCreation.ArrayInitializer.InitializerElements;
+            for (int i = 0; i < elements.Count; i++)
+            {
+                pattern.Append($"new $0[] ${i + 1}");
+                args.Add(elements[i]);
+                pattern.Append(",");
+            }
+
+            pattern.Remove(pattern.Length - 1, 1);
+            pattern.Append("}");
+
+            var newCreation = factory.CreateExpression(pattern.ToString(), args.ToArray());
+            currentCreation.ReplaceBy(newCreation);
+        }
+
+        private void ConvertUsages(CSharpElementFactory factory)
+        {
+            foreach (var usage in myTreeNodePointers)
+            {
+                var node = usage.GetTreeNode() as ICSharpExpression;
+                var elementAccessExpression = ElementAccessExpressionNavigator.GetByOperand(node as ICSharpExpression)
+                    .NotNull("ElementAccessExpressionNavigator.GetByOperand(node as ICSharpExpression) != null");
+                var arguments = elementAccessExpression.Arguments;
+                if (node == null || arguments.Count < 2)
+                    continue;
+
+                var newUsage = factory.CreateExpression("$0[$1][$2]", node, arguments[0], arguments[1]);
+                elementAccessExpression.ReplaceBy(newUsage);
+            }
+
+        }
+
+        private void ConvertArrayCreationWithoutInitializer(CSharpElementFactory factory)
+        {
+            var currentCreation = myArrayCreationExpression;
+
+            var firstSize = currentCreation.Sizes[0].NotNull("currentCreation.Sizes[0] != null").CopyWithResolve();
+            var secondSize = currentCreation.Sizes[1].NotNull("currentCreation.Sizes[1] != null").CopyWithResolve();
```

Comment:
Does this only work with multidimensional arrays of rank 2? If so, that should either be part of the warning, or disable the quick fix for larger ranks

Reply:
Quick fix is disabled for larger ranks

Reply:
Ah, missed that. Thanks